Jetpack connect: Add and use from selector

sirreal opened the pull request:
This PR adds a selector to the `from` data included in Jetpack Connect authorization.

It also leverages `from` in the AuthLoggedIn component.

It moves `isWoo` and `isSSO` out of the parent and into local methods because they also depend on the data from this selector and are unused in the parent.

### Testing
1. No behavioral changes.
1. Ensure the Woo, SSO, and JPO flows all continue to work.
1. Normal flows should continue to work.

sirreal committed:
Add getJetpackConnectFrom selector

sirreal committed:
Use getJetpackConnectFrom in auth-logged-in-form

sirreal committed:
Remove unused prop from propTypes

sirreal committed:
Move isWoo to consuming component, use from selector

sirreal committed:
Add radix to parseInt in selector

sirreal committed:
Move isSSO to authLoggedInForm

sirreal committed:
Use already selected queryDataSiteId

sirreal committed:
Update snapshot

sirreal commented:
I'd appreciate if SSO, JPO, and Woo behavior verified by @ebinnion and @v18 🙇

seear commented on client/state/jetpack-connect/selectors.js:
```diff
@@ -170,7 +170,7 @@ const getGlobalSelectedPlan = function( state ) {
 	const authorizationData = getAuthorizationData( state );
 	if ( authorizationData.queryObject && authorizationData.queryObject.client_id ) {
-		return parseInt( authorizationData.queryObject.client_id );
+		return parseInt( authorizationData.queryObject.client_id, 10 );
```
10?

sirreal replied:
Base-10

> `radix`
> An integer between 2 and 36 that represents the _radix_ (the base in mathematical numeral systems) of the above mentioned string. Specify `10` for the decimal numeral system commonly used by humans. **Always specify this parameter** to eliminate reader confusion and to guarantee predictable behavior. Different implementations produce different results when a radix is not specified, usually defaulting the value to 10.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/parseInt

seear replied:
> Always specify this parameter to eliminate reader confusion

😄

seear commented on client/jetpack-connect/auth-logged-in-form.jsx:
```diff
@@ -137,7 +135,7 @@ class LoggedInForm extends Component {
 		// For SSO, WooCommerce Services, and JPO users, do not display plans page
 		// Instead, redirect back to admin as soon as we're connected
-		if ( nextProps.isSSO || nextProps.isWoo || this.isFromJpo( nextProps ) ) {
+		if ( nextProps.isSso( nextProps ) || this.isWoo( nextProps ) || this.isFromJpo( nextProps ) ) {
```
Should this be `this.isSso( nextProps )` instead of `nextProps.Sso( nextProps )`?

sirreal replied:
Yes!

sirreal replied:
Fixed in 38112cf107c9d7cfe33a338aadab532c61015519, thanks!

sirreal committed:
Fix nextProps should be this

seear commented:
Trying to test the SSO flow but I don't quite understand it. How can SSO be enabled if the site is not yet connected. Or is this form used for part of SSO not related to connecting the site?

ebinnion commented:
With SSO, we auto-authorize secondary users after they've approved SSO.

So, your will already be connected to WordPress.com and you'll need to have the SSO module on. Then, with a test account, you'll need to SSO in. After approving SSO, you should be auto-authorized.

seear commented:
> With SSO, we auto-authorize secondary users after they've approved SSO.

Thanks @ebinnion!

ebinnion commented:
@sirreal - something is broken in the SSO auto-authorize flow. When I test against master, my secondary test account is auto-authorized.

But, with this branch, it's asking me to approve the connection after I clicked the log in button for SSO.

sirreal commented:
Thanks @ebinnion! I'll track down the issue…

ebinnion commented on client/jetpack-connect/auth-logged-in-form.jsx:
```diff
@@ -118,7 +116,7 @@ class LoggedInForm extends Component {
 		// isSSO is a separate case from the rest since we have already validated
 		// it in authorize-form.jsx. Therefore, if it's set, just authorize and redirect.
-		if ( this.props.isSSO || doAutoAuthorize ) {
+		if ( this.isSso() || doAutoAuthorize ) {
```
We're not passing props to `isSso` here. I wonder if that's the issue? I'm not super familiar with the operator we're using in the `isSso` method below.

sirreal replied:
Maybe being too fancy there 😓 

https://github.com/Automattic/wp-calypso/pull/19956/files#diff-e8644e2359a41377a490f6cbe8003db3R186

```
isSso( { from, queryDataSiteId } = this.props )
```

* `{ from, queryDataSiteId }` -> destructuring assignment
* `... = this.props` -> default argument

We should be assigning `from` and `queryDataSiteId` from the properties on the first argument supplied, and if that argument is not supplied (`undefinted`), then we should default to using `this.props` as the argument.

In theory this should work, but I'll need to verify.

sirreal committed:
Use queryData selector

sirreal committed:
Add isSso tests

sirreal committed:
Fix isSso logic

With the move to use the selector, a strict equals check was failing.
Improve the test robustness.

sirreal commented:
@ebinnion That was a bug introduced by moving from directly accessing the state to using the selector which returns an int, Strict equality check was failing, thanks for catching it. 🙂 

I've added tests and fixed the problem 🎉 

---

@v18 I'm attempting to test the Woo flows but I'm not sure how to hit auth on my dev environment. Can you provide testing instructions or test this please?

ebinnion commented:
SSO tests well for me now. 👍

sirreal commented:
Rebased to fix conflict.

---

@Automattic/hydra could you test or provide testing instructions for how to verify this change plays well with the WooCommerce connection flow?

sirreal committed:
Use this.props default argument for shouldRedirJPS

sirreal commented:
Rebased again

sirreal commented on client/jetpack-connect/auth-logged-in-form.jsx:
```diff
@@ -172,32 +172,46 @@ class LoggedInForm extends Component {
+	}
+
+	isWoo( { from } = this.props ) {
+		return includes( [ 'woocommerce-setup-wizard', 'woocommerce-services' ], from );
```
@marcinbot Thanks for the feedback (https://github.com/Automattic/wp-calypso/pull/19956#issuecomment-347905765).

Are you saying that this string should be changed from `woocommerce-services` to `woocommerce-services-auto-authorize` ?

That string is not introduced in this PR, so that flow isn't working in production 🤔

sirreal commented:
I've tested the Woo wizard according to instructions in https://github.com/Automattic/woocommerce-services/pull/1114 and it behaves as expected.

sirreal committed:
Add missing propTypes
